### ISSUE TITLE
Structured Citations

### ISSUE THREAD

**Issue by GaziYucel** [pkp/pkp-lib#10692]
Update 01.10.2025

Still to finish:
- [ ] test native import export
- [ ] adapt article_details.tpl for each theme
- [ ] adapt jatsTemplate
- [ ] adapt crossrefReferenceLinking 

<hr>

### Overview
The structured citations feature in OJS is designed to streamline the process of managing references within submissions. It allows authors and editors to input, process, and manage citations efficiently, ensuring accurate metadata retrieval and integration.

### Acceptance criteria and detailed design specifications

**`User Path 1: Enable Structured Citations Setting at the Journal Level`**

- A toggle is available under Settings > Workflow > Submission > Metadata to enable or disable Structured Citations by default. 
- However, this setting is not mandatory—editors can override it for individual submissions by going to the Publication > References tab and adjusting the setting there 

**`User Path 2: Structured Citations during Article Submission Process`**

- When a new submission is started by the author or editor but has not yet been submitted, the References field will appear as a simple text input with no structured citations.
- Once the submission is officially submitted to the journal, the Structured Citations feature becomes available in the Publication tab and can be accessed from all workflow stages.

**`User Path 3: Author or Editor Adds Citations During Workflow`**

**Submission > References Tab**
- A dedicated References section allows authors (or editors) to input references as free-text, with each reference entered on a new line.

- A brief instruction appears above the field:

> “Enter each reference on a new line. You can paste one or multiple references at a time. Click ‘Add’ to process them into structured citations.”

**"Add" Button Behavior**
- Clicking Add triggers metadata lookup via external services (e.g., Crossref) and attempts to match the reference with available DOI and citation metadata.
- While processing, a status message appears:

> “Processing references… 0 of 4 added so far. This may take a few moments. You can continue working while the references are being retrieved.”

- A caution is also displayed:

> “Some references may not return metadata. You can edit them manually once they’re added to the list.”

**`User Path 4: Display of Processed References in Structured Format`**

- Once processed, structured references appear in a tabular list below the text box.
- A success message is displayed:

> “All references successfully added. You can review, edit, or remove them at any time.”

**Table Actions**

Each reference row includes:
- Formatted citation (retrieved or entered manually)
- DOI (if matched)
- Action buttons: Edit | Delete
- Clicking on "Edit" will open a modal where the user can fill more information about the citations or modify the existing one to make it structured

If users paste new references into the textbox and click Add again:

- The system detects and adds only unique entries.
- Duplicate references are ignored.

**`User Path 5: Deleting References`**

- A "Delete All Structured Citations" button appears above the table when entries exist.
- Clicking it opens a confirmation modal with the following message:
> “Are you sure you want to delete all structured citations? This will remove all references currently listed. You’ll need to re-enter and process your references again if you continue.”
- Options: Cancel | Delete All

### 📽 Video Walkthrough:

https://youtu.be/KYPBl66oxQc
Figma Prototype - https://www.figma.com/proto/Wf7sDlUg2372jaKKTJ0Mgz/OJS-3.4-3.5?page-id=7200%3A7069&node-id=13467-15870&viewport=1476%2C-596%2C0.16&t=QyYooCIJnrzZccgk-1&scaling=min-zoom&content-scaling=fixed&starting-point-node-id=13467%3A15870
Figma Design Link - https://www.figma.com/design/Wf7sDlUg2372jaKKTJ0Mgz/OJS-3.4-3.5?node-id=8934-12758&t=mSiC57aKAv6sL8eW-4 

### Additional Information from Backend Perspective 

**Add structured citations feature to OJS 3.5 and above.** 

Focus will be structuring references which are input in the free input field on the References section in the backend UI. The location of the structured citations will be under the current References text area field. 

- database
  - citations: citation_id, publication_id, raw_citation, seq
  - citation_settings: citation_setting_id, citation_id, locale, setting_name, setting_value
  - authors and other fields in the schema are an array of objects, these will be saved as a json blob in citation_settings table
  - publication: citationsMetadataLookup
- UI Library components will be used
- the location in the UI will be under the References text area. 
- adding references will still be done by adding in the References textarea as done now
- saving references will add a job to: 
  -  extract pids
  - retrieve structured meta data from external services
- all citations with an doi will be processed automatically: 
  - extract doi
  - if doi found: structured citations will be retrieved from openalex
  - if doi not found, retrieve from crossref api:
    - match with score, title, family names authors
- when publication is published, a job will be added to deposit the citations to Open Citations repository
- include citations in exports
- citations will not be multi-lingual in this version

**Pull requests**

Part 2: Structured citations (against main)
- https://github.com/pkp/pkp-lib/pull/11884
- https://github.com/pkp/ui-library/pull/716
- https://github.com/pkp/ojs/pull/5115
- https://github.com/pkp/omp/pull/2130
- https://github.com/pkp/ops/pull/1107

Part 1: [Prepare DB for structured citations to come #11238](https://github.com/pkp/pkp-lib/issues/11238)
- https://github.com/pkp/pkp-lib/pull/11239
- https://github.com/pkp/ojs/pull/4819
- https://github.com/pkp/omp/pull/1947
- https://github.com/pkp/ops/pull/948

PRs from @GaziYucel, that were the bases for the PRs "Part 2" above:

Part 2: structured citations (against main)
- pkp-lib: https://github.com/pkp/pkp-lib/pull/11830
- ui-library: https://github.com/pkp/ui-library/pull/704
- ojs: https://github.com/pkp/ojs/pull/5098
- omp: https://github.com/pkp/omp/pull/2123
- ops: https://github.com/pkp/ops/pull/1095

Part 3: open citations
- https://github.com/GaziYucel/openCitations

Structured citations (against stable-3_5_0)
- ~https://github.com/pkp/pkp-lib/pull/11270~
- ~https://github.com/pkp/ui-library/pull/600~
- ~https://github.com/pkp/ojs/pull/4839~
- ~https://github.com/pkp/omp/pull/1959~
- ~https://github.com/pkp/ops/pull/961~
- ~https://github.com/pkp/crossrefReferenceLinking/pull/32~
- ~https://github.com/pkp/googleScholar/pull/30~ [no changes]
- ~https://github.com/pkp/jatsTemplate/pull/87~

**Repository branches**
- https://github.com/GaziYucel/ojs/commits/I10692-structured-citations/ 
- https://github.com/GaziYucel/omp/tree/I10692-structured-citations
- https://github.com/GaziYucel/ops/commits/I10692-structured-citations/
- https://github.com/GaziYucel/pkp-lib/commits/I10692-structured-citations/
- https://github.com/GaziYucel/ui-library/commits/I10692-structured-citations/
- https://github.com/GaziYucel/googleScholar/commits/I10692-structured-citations/
- https://github.com/GaziYucel/jatsTemplate/commits/I10692-structured-citations/
- https://github.com/GaziYucel/crossrefReferenceLinking/commits/I10692-structured-citations/
- https://github.com/GaziYucel/opencitations

**Backend UI**

~https://www.youtube.com/watch?v=yWpQ1GNexow~

**Schema Citation**
- id: id of citation
- publicationId: id of publication
- seq: position or sequence of citation in citation list
- authors: array of objects { displayName, givenName, familyName, orcid, wikidata, openalex }
- doi: The DOI itself, such as 10.1234/5a6b-7c8d.
- url: URL for the work.
- urn: URN for the work.
- arxiv: Arxiv id / url.
- handle: Handle id / url.
- openAlex: OpenAlex id / url.
- wikidata: Wikidata qid / url.
- rawCitation: citation in raw form
- title: The title of the work.
- date: The publication date, formatted as an ISO 8601 date e.g. 2018-02-13.
- type: The type or genre of the work, e.g. journal-article.
- volume: The volume of the issue of the journal, e.g. 495.
- issue: The issue of the journal, e.g. 7442.
- pages: Number of pages, e.g 64.
- firstPage: The first page of the work/article, e.g. 49.
- lastPage: The last page of the work/article, e.g. 59.
- sourceName: Name of the source, e.g. journal name
- sourceIssn: Issn_l of the source, e.g. journal issn_l
- sourceHost: Publisher name of the host, e.g. publisher name
- sourceType: Type of the source, e.g. journal, book
- isStructured: Whether this citation is structured or not.
- isProcessed: Whether this citation si processed.

***Schema Publication***
- citationsMetadataLookup: Whether or not to use the structured citations.

**Comment by Tribunal33** [comment]
If possible could we put some sample references with expected outcomes in the main requirements section for testing purposes?  Don't know if something like that is already going to be an automated cypress test but would be a good candidate.

**Comment by asmecher** [comment]
Moving this off to 3.5.0-x -- the infrastructural accommodations have been merged for 3.5.0-0, but the feature needs ongoing work.

**Comment by bozana** [comment]
hi @Tribunal33, everything major is merged, so it is all yours now :-)

References need to be enabled under settings > workflow > metadata > references. There, when enabled, one can select the default setting for every new publication, if metadata lookup is per default enabled or disabled. On every workflow publication page, on the references page, it can then be determined if the metadata lookup is enabled or disabled.
If not enabled, everything should work as till now.

For structured citations:
If you take a citation that has a DOI, e.g. 10.3897/rio.4.e31656, i.e. the citation:
`Hauschke C, Cartellieri S, Heller L (2018) Reference implementation for open scientometric indicators (ROSI). Research Ideas and Outcomes 4: e31656. https://doi.org/10.3897/rio.4.e31656`
then you can see what data we get from OpenAlex like this: https://api.openalex.org/works/doi:https://doi.org/10.3897/rio.4.e31656.
Here is what data we use and how we map them to citations settings in OJS/OMP/OPS: https://github.com/pkp/pkp-lib/blob/main/classes/citation/externalServices/openAlex/Mapping.php#L35-L48.

If you leave the DOI or use a citation that has no DOI, first Crossref will be checked, to see if we can find a DOI, that is then used for OpenAlex for metadata lookup, to get the citation data (structured citation).

The processing (metadata lookup) happens in jobs - a job per citation - so eventually it can take a while, depending on how your jobs running is set up.

The citations can be edited, and not structured citations re-processed.

I hope I have mentioned everything major 🤔 

Thanks a lot!

**Comment by kaitlinnewson** [comment]
@GaziYucel One minor issue I'm noticing is the processing of arxiv ids, as its enforcing a URL structure but it should not be - here's an example of citations using arxiv that I was using: https://www.underleaf.ai/blog/how-to-cite-arxiv-papers

**Comment by Tribunal33** [comment]
@bozana, after my first round of testing, here are some things I'm not sure on from examining the figma designs, demo video and current Acceptance criteria.
- [ ] Author should be able to add citations but currently it gives an error after selecting add
<img width="406" height="188" alt="Image" src="https://github.com/user-attachments/assets/5c5594f9-28af-4885-b8ba-aa63c4c485c6" />

- [ ] Author should not be able to toggle the metadata lookup checkbox it should be disabled.  Currently there is an error message after trying to complete the process.  From the video this was to be disabled.
<img width="447" height="268" alt="Image" src="https://github.com/user-attachments/assets/f1cf43e0-5b59-4040-a912-d4e18af2e951" />

- [ ] Author is not able to delete all citations.  This is disabled.  I believe they should be allowed to delete all?
<img width="769" height="115" alt="Image" src="https://github.com/user-attachments/assets/ad9dcf98-3fb5-4fad-9dc0-775ffbfdd749" />

- [x] ~~Not seeing ability to have `wikidata` or `opendata` reference source bubbles as indicated in figma designs~~ This is removed from scope.
- [ ] As @kaitlinnewson pointed out the field for Arxiv should not enforce URL validation restrictions as the correct way to source it is something like this "arXiv preprint arXiv:1706.03762".  I'm not sure if this needs to be fixed as it isn't specified so perhaps it is only accepting urls similar to the way DOIs would need to be a url.  Example of arxiv url: https://arxiv.org/abs/1706.03762
<img width="325" height="148" alt="Image" src="https://github.com/user-attachments/assets/924dac00-907a-4dca-b3e4-1a40ea5ceca5" />

- [ ] Not sure how functional it is to "continue working" when processing large batches of citations.  I tried doing 100 citations and then tried navigating the site. It essentally doesn't work until they are processed.  I, also, tried doing a 100 citations as one users then popping over to another user in different browser and tried to navigate the site.  Seems to be affected there too.  I'm not sure the jobs process is doing async correctly.   

**Things to improve (Could open up another part 2)**
- [ ] Should limit to 100 citations at a time (for both security and navigating the site while it is processing is sluggish)
- [x] ~~Publication date should be a date picker~~ (added)
- [ ] Arxiv does not need URL validation - arXiv:2401.12345 is valid.
- [ ] Increase the size of the citation input field.  Feels too small for the amount of information and spacing between citations.
- [ ] Disable Metadata lookup if it is enabled site-wide in settings for author.
- [ ] Editing citation should have more informative validation warnings for Type and source type or have auto complete or dropdown list
- [ ] Editing citation should limit the characters for free text fields.
- [ ] Async functionality with jobs processing.

**Comment by bozana** [comment]
I thought I have commented "wikidata and opendata bubbles" here, but I do not find my comment any more 😅 
We had a long discussion about this in a Mattermost chanel with Devika and Gazi. Here is the last from Devika:

> According to me, I think we can add the links to take us to the source data it will be cool to have. But it should take us to legitimate places and not just our repository of these references. Thats what I feel. And if they are links I would change the design a bit to make it buttons and not tags. I thought of using them as tags cause it also indicates what sources picked up. Maybe I can do both.

Thus, because only OpenAlex is the source of our data, we only display OpenAlex bubble link.

**Comment by Tribunal33** [comment]
Sorry @bozana I was following what I saw in the figma designs.  I'm not sure, however, what those tags/bubbles are supposed to represent.  But as I understand it, we are only using OpenAlex for now.  I can see OpenAlex and that's what is in the A/C so I pass it 😄 .  It doesn't say when this bubble should show up or what the bubble should link to so I will assume it is working correctly.  

<img width="845" height="217" alt="Image" src="https://github.com/user-attachments/assets/461015a4-a967-413a-a877-bbd62fb0f051" />

**Comment by Tribunal33** [comment]
@GaziYucel and @bozana I see that the source type and type fields have added additional controlled vocabulary.  This is addressed in part 2 of using a dropdown or autocomplete feature so the user can pick from the list.  I am currently seeing an issue with the SourceType as it isn't accepting two word entries such as "ebook platform" or "book series".  Additionally I'm seeing validation for approved Type such as "libguides" and "letter".  I didn't check them all so not sure to extent others are missing.  

<img width="741" height="602" alt="Image" src="https://github.com/user-attachments/assets/adaea55e-9bc2-45c1-ac0e-f07ac48fae65" />

**Comment by GaziYucel** [comment]
Hi @Tribunal33 I am currently addressing the sourceType and Type. These two will be drop downs, working locally but not pushed yet. I will let you know when I am finished. Thanks.

**Comment by bozana** [comment]
Because all remaining tasks were moved to https://github.com/pkp/pkp-lib/issues/11902, this issue could be closed, right @Tribunal33 ?

**Comment by Tribunal33** [comment]
@bozana not all issues were moved.  The key ones for this are author permissions.  Currently Author cannot submit citations after submission is complete.  They are also unable to delete all citations.  This is part of A/C and needs to be fixed before we can close.

Also the source type and type validation issue, but If you want to merge that with my suggested dropdown list for these fields into 11902 that is fine by me.

**Comment by bozana** [comment]
Hi @Tribunal33, I added the author issues there, type and source type were already mentioned there. So maybe to close this issue now?
Gazi is preparing all the fixes in one PR, we can then assign it to the new issue https://github.com/pkp/pkp-lib/issues/11902

**Comment by Tribunal33** [comment]
I'm moving the expectations of the controlled vocab to #11902 as I think the enhancement to the fields will solve both problems.  This passes QA and I will close now.  

Here are the testing scenarios for this scope.  Some might not apply but part of initial test plan creation (like bubble links)

1. Test setting to Enable Structured Citations enables structured citations section
2. Test setting to Enable metadata lookup can be toggled on or off for the journal
3. Test setting to require citations during submission will appear as structured citations
4. Test adding repeated unique single citations
5. Test Delete all citations button
6. Test adding multiple citations
7. Test adding duplicated citations
8. Test deleting a single citation from the 3 dots menu
9. Test editing a single citation from the 3 dots menu
10. Test expand all and collapse all toggle works
11. Test adding 50+ citations and continue to navigate the site while it processes.
12. Test citation indication states.
13. Test Edit citation validation
14. Test Edit citation SourceType and Type validation
15. Test Edit citation field validation for Arxiv
16. Test Edit citation field date is a date picker
17. Test Search citations field
18. Test disabling Metadata look up toggle after structured citations have been added
19. Test enabling Metadata look up toggle after strucutre citations have been added
20. Test specific citation for wikidata
21. Test specific citation for Opendata
22. Test specific citation for OpenAlex
23. Test Wikidata citation bubble link
24. Test Opendata citation bubble link
25. Test OpenAlex citation bubble link
26. Test Author can add citations
27. Test Author can delete all citations
28. Test Author can not toggle enable metadata lookup when enabled sitewide

Author specific ones will be part of 11902.
